### PR TITLE
[Locator] [WIP] Allow new folder creation within the locator

### PR DESCRIPTION
**Pull request by BogdanAlexandru**
This is a WIP PR for #579

## Todo List:
- [x] Auto-scroll to new folder
- [ ] New tests
- [ ] Layout work
- [ ] Getting rid of the form validation issue
- [x] Validate new folder name somehow (same rules as when creating a new folder)
- [ ] Make the behavior between objects that are editable on creation and those that are not be consistent
- [ ] Freeze/unfreeze the dialog while the new folder is being created
- [ ] Anything else that comes up during review

---------------

Went through a busier period so progress was slow, should go faster now 👍 

Asking for thoughts on my approach and on anything that would make things go smoother.

My idea was to allow the locator to create new folders when requested, but there's a bunch of things that seem to not welcome my approach.

img

Aside from that, I'm trying to take out the whole internal folder creation logic and bring it to the locator, duplicating it. I can't reuse much of the creation + saving logic because it's tied to the UI (see actions), and an extra panel cannot be spawned on top of the existing one - I haven't checked but I think it would also cause trouble when it comes to nested transactions, if this scenario was not expected initially.

Do you see any way in which I could handle this without causing changes to the other components (such as form validation - functionality that tells the form not to try and validate some specific fields; or creation + saving actions)?

My gut feeling tells me the quickest & easiest goal to try and reach would be to allow an actual new folder creation (with the whole UI workflow), so clicking New Folder would start a regular creation process (i.e. `new CreateAction(folderType, selectedParent, context).perform()`), but that goes way against the specs in the issue.

Cleanest I think would be taking the UI bits out of the creation & saving logic, which would then take in `creationInfo` as a param, indicating what properties the new object would have. The locator would then do something like

```
var info = { name: enteredName };
new CreateAction(folderType, parent, info).perform();
```

That seems like a major change wrt how everything works though.

### CODE REVIEW

File: platform/commonUI/edit/src/creation/LocatorController.js [hunk 1]
```diff
@@ -81,11 +84,46 @@ define(
                         !!$scope.treeModel.selectedObject
                     );
                 }
+
+                $scope.canCreateNewFolder = ($scope.treeModel.selectedObject.model.type === "folder");
```
Comment (akhenry) [inline platform/commonUI/edit/src/creation/LocatorController.js]: We should probably use the 'composition' policy here. eg
```
policyService.allow(
    "composition",
    parentType,
    childType
)
```

File: platform/commonUI/edit/src/creation/LocatorController.js [hunk 2]
```diff
@@ -81,11 +84,46 @@ define(
...
             }
 
+            function createNewFolder(name, parent) {
+                var folderType = typeService.listTypes().filter(function (type) {
```
Comment (akhenry) [inline platform/commonUI/edit/src/creation/LocatorController.js]: Could use `typeService.getType()` here I think

**Comment by akhenry** [comment]
@BogdanAlexandru I think you're taking basically the right approach. I've made two specific comments on details of API usage, but I think you're on the right track. Trying to leverage CreateAction or SaveAsAction for the new folder is probably doomed to failure. There should not be too much logic that we need to build into the `LocatorController` here. I've tried to capture it below, but I may have missed something.

* We need to check whether the object currently selected in the container can itself contain a folder which we can use a policy for (see my specific comment on this). 
* Instantiation can be used for creating the new folder object (which you're already doing), and making a call to `PersistenceCapability.persist()` on the new folder will mark it dirty on the transaction, so it will be persisted along with the new object. 
* You'll need to make sure that the new folder is scrolled to, and automatically selected in the locator. If it's selected, this should be enough to ensure that the new object has its location set to the newly created folder object.

Let me get back to you on the form validation issue.

**Comment by charlesh88** [comment]
Cool!

- Charles

On Dec 19, 2016, at 12:47 PM, Andrew Henry <notifications@github.com<mailto:notifications@github.com>> wrote:


@BogdanAlexandru<https://github.com/BogdanAlexandru> I think you're taking basically the right approach. I've made two specific comments on details of API usage, but I think you're on the right track. Trying to leverage CreateAction or SaveAsAction for the new folder is probably doomed to failure. There should not be too much logic that we need to build into the LocatorController here. I've tried to capture it below, but I may have missed something.

  *   We need to check whether the object currently selected in the container can itself contain a folder which we can use a policy for (see my specific comment on this).
  *   Instantiation can be used for creating the new folder object (which you're already doing), and making a call to PersistenceCapability.persist() on the new folder will mark it dirty on the transaction, so it will be persisted along with the new object.
  *   You'll need to make sure that the new folder is scrolled to, and automatically selected in the locator. If it's selected, this should be enough to ensure that the new object has its location set to the newly created folder object.

Let me get back to you on the form validation issue.

**Comment by BogdanAlexandru** [comment]
Quick demo with the functionality so far. The cursor seems to disappear once I click within the window, sorry about that:
Imgur

Thing still left to be done:

- [ ] Auto-scroll (right now it's auto-select, but the tree doesn't scroll to the selected element necessarily)
- [ ] New tests
- [ ] Layout work
- [ ] Getting rid of the form validation issue
- [ ] Validate new folder name somehow (same rules as when creating a new folder)
- [ ] Anything else that comes up during review

@akhenry Simply instantiating the new folder didn't result in it being created, I also had to compose its parent with it, and persist both the new folder and its parent (visible in my latest changes). Is this correct, or do you see any flaws in my approach?

**Comment by BogdanAlexandru** [comment]
For item

- [ ] Freeze/unfreeze the dialog while the new folder is being created

I thought a simple method to prevent interaction while the new folder is being created would be to show a blocking dialog as can be seen when performing Save. This would avoid anything voodoo that involves manipulating the rest of the dialog from the locator. The dialog service is not allowed to spawn 2 dialogs however.

How much of a stretch would it be to come up with a "blocking info" dialog concept that doesn't care about the one-dialog limit (not sure how much work that would be, just removing the check for any existing dialogs when attempting to show the blocking dialog doesn't order the overlays properly)? Its sole purpose would be to block interactions while something is in progress, always shown above all other overlays.

Or if there's anything else that can be done and doesn't touch on what looks like an important functional rule, I'm all for it. 

Let me know.

**Comment by akhenry** [comment]
> @akhenry Simply instantiating the new folder didn't result in it being created, I also had to compose its parent with it, and persist both the new folder and its parent (visible in my latest changes). Is this correct, or do you see any flaws in my approach?

@BogdanAlexandru Ah, yep, you'll need to add it to the parent's composition. You'll also want to set the `location` property on the new folder's model to be the ID of the parent, otherwise the new object will appear as a link in the UI. The act of mutating both the parent and the child *should* automatically mark them dirty on the transaction, without needing to explicitly call `.persist()` on either object. They will then be persisted along with the new object.

> There is a difference in behavior between elements that have an edit mode started on creation and those that don't.

Gah! That will be because of the [`TransactingMutationListener`](https://github.com/nasa/openmct/blob/6fa5a312175545a0a7f44bc2339849ce300d7b58/platform/core/src/runs/TransactingMutationListener.js#L53). This class listens for object mutation and marks them dirty if a transaction is under way. If not, it will automatically persist the thing being mutated. So the trick here is to explicitly start a transaction, because we obviously don't want anything being persisted until the user presses `save`. The easiest way to start a transaction is to put the parent object into edit mode programmatically. You'll need to do this conditionally though, trying to put an object into edit mode that is already being edited will throw an error. You could do something like:

```javascript
domainObject.getCapability("editor").edit()
domainObject.getCapability("editor").isEditContextRoot()
```

**Pull request closed without merging**